lib/MongoDB/ChangeStream.pm

```diff
@@ -0,0 +1,201 @@
+    is => 'ro',
+    isa => MongoDBCollection,
+    required => 1,
+);
+
+has options => (
+    is => 'ro',
+    isa => HashRef,
+);
+
+has pipeline => (
+    is => 'ro',
+    isa => ArrayOfHashRef,
+    required => 1,
+);
+
+has full_document => (
+    is => 'ro',
+    isa => Str,
+    predicate => '_has_full_document',
+);
+
+has _resume_token => (
+    is => 'rw',
+    init_arg => 'resume_after',
+    predicate => '_has_resume_token',
+    lazy => 1,
+    builder => '_build_resume_token',
+);
+
+sub BUILD {
+    my ($self) = @_;
+
+    # starting point is construction time instead of first next call
+    $self->_cursor;
+}
+
+sub _build_cursor {
+    my ($self) = @_;
+
+    my $pipeline = $self->pipeline;
+
+    my @pipeline = @{ $self->pipeline || [] };
+    @pipeline = (
+        {'$changeStream' => {
+            ($self->_has_full_document
+                ? (fullDocument => $self->full_document)
+                : ()
+            ),
+            ($self->_has_resume_token
+                ? (resumeAfter => $self->_resume_token)
+                : ()
+            ),
+        }},
+        @pipeline,
+    );
+
+    return $self->collection->aggregate(
+        \@pipeline,
+        {
+            %{ $self->options || {} },
+            cursorType => 'tailable_await',
+        },
+    );
+}
+
+=head1 STREAM METHODS
+
+=cut
+
+=head2 next
+
+    $change_stream = $collection->watch(...);
+    $change = $change_stream->next;
+
+Waits for the next change in the collection and returns it.
+
+B<Note>: This method will wait for the amount of milliseconds passed
+as C<maxAwaitTimeMS> o L<MongoDB::Collection/watch> or the default. It
+will not wait indefinitely.
+
+=cut
+
+sub next {
+    my ($self) = @_;
+
+    my $change;
+    my $retried;
+    while (1) {
+        last if try {
+            $change = $self->_cursor->next;
+            1 # successfully fetched result
+        }
+        catch {
+            my $error = $_;
+            if (
+                not($retried)
+                and $error->$_isa('MongoDB::Error')
+                and $error->_is_resumable
+            ) {
+                $retried = 1;
+                $self->_cursor($self->_build_cursor);
+            }
+            else {
+                die $error;
+            }
+            0 # failed, cursor was rebuilt
+        };
+    }
+
+    # this differs from drivers that block indefinitely. we have to
+    # deal with the situation where no results are available.
+    if (not defined $change) {
+        return;
+    }
+
+    if (exists $change->{_id}) {
+        my $resume_token = $change->{_id};
+        $self->_resume_token($resume_token);
+        return $change;
+    }
+    else {
+        MongoDB::InvalidOperationError->throw(
+            "Cannot provide resume functionality when the ".
+            "resume token is missing");
+    }
+}
+
+1;
+
+=head1 SYNOPSIS
+
+    $stream = $collection->watch( $pipeline, $options );
+    while (my $change = $stream->next) {
```

Comment:
Please wrap this in an infinite loop with a comment about how the inner loop exits when there are no changes available.